sklearn/utils/estimator_checks.py

```diff
@@ -395,57 +402,138 @@ def _yield_all_checks(estimator, legacy: bool):
+        Check to be marked.
+    mark : "xfail" or "skip" or None
+        Whether to mark the check as xfail or skip.
+    pytest : pytest object, default=None
+        Pytest module to use to mark the check.
+    """
+    should_be_marked, reason = _should_be_skipped_or_marked(
+        estimator, check, expected_failed_checks
+    )
+    if not should_be_marked or mark is None:
         return estimator, check
-    else:
+
+    estimator_name = estimator.__class__.__name__
+    if mark == "xfail":
         return pytest.param(estimator, check, marks=pytest.mark.xfail(reason=reason))
+    else:
 
+        @wraps(check)
+        def wrapped(*args, **kwargs):
+            raise SkipTest(
+                f"Skipping {_check_name(check)} for {estimator_name}: {reason}"
+            )
 
-def _maybe_skip(estimator, check):
-    # Wrap a check so that it's skipped if needed (see conditions in
-    # _should_be_skipped_or_marked())
-    # This is similar to _maybe_mark_xfail(), but this one is used by
-    # check_estimator() instead of @parametrize_with_checks which requires
-    # pytest
-    should_be_skipped, reason = _should_be_skipped_or_marked(estimator, check)
-    if not should_be_skipped:
-        return check
+        return wrapped
 
-    check_name = check.func.__name__ if isinstance(check, partial) else check.__name__
 
-    @wraps(check)
-    def wrapped(*args, **kwargs):
-        raise SkipTest(
-            f"Skipping {check_name} for {estimator.__class__.__name__}: {reason}"
-        )
+def _should_be_skipped_or_marked(
+    estimator, check, expected_failed_checks: dict[str, str] | None = None
+) -> tuple[bool, str]:
+    """Check whether a check should be skipped or marked as xfail.
+
+    Parameters
+    ----------
+    estimator : estimator object
+        Estimator instance for which to generate checks.
+    check : partial or callable
+        Check to be marked.
+    expected_failed_checks : dict[str, str], default=None
+        Dictionary of the form {check_name: reason} for checks that are expected to
+        fail.
 
-    return wrapped
+    Returns
+    -------
+    should_be_marked : bool
+        Whether the check should be marked as xfail or skipped.
+    reason : str
+        Reason for skipping the check.
+    """
+
+    expected_failed_checks = expected_failed_checks or {}
+
+    check_name = _check_name(check)
+    if check_name in expected_failed_checks:
+        return True, expected_failed_checks[check_name]
 
+    return False, "Check is not expected to fail"
 
-def _should_be_skipped_or_marked(estimator, check):
-    # Return whether a check should be skipped (when using check_estimator())
-    # or marked as XFAIL (when using @parametrize_with_checks()), along with a
-    # reason.
-    # Currently, a check should be skipped or marked if
-    # the check is in the _xfail_checks tag of the estimator
 
-    check_name = check.func.__name__ if isinstance(check, partial) else check.__name__
+def checks_generator(
```

Comment:
`check_estimator(..., generate_only=True)` is now deprecated, and this function is the replacement.

Reply:
Maybe rename to `yield_estimator_checks`?

Reply:
Or `generate_estimator_checks`?

Reply:
Or `estimator_checks_generator`?

Reply:
Used the last one.